src/raw_node.rs

```diff
@@ -344,92 +380,259 @@ impl<T: Storage> RawNode<T> {
-        if has_unapplied_entries {
+
+        if self.snap().map_or(false, |s| !s.is_empty()) {
             return true;
         }
-        if raft.soft_state() != self.prev_ss {
+
+        if raft
+            .raft_log
+            .has_next_entries_between(self.commit_since_index, self.last_persisted_index)
+        {
             return true;
         }
-        let hs = raft.hard_state();
-        if hs != HardState::default() && hs != self.prev_hs {
+
+        if !raft.msgs.is_empty() || !self.messages.is_empty() {
             return true;
         }
         false
     }
 
-    /// HasReady called when RawNode user need to check if any Ready pending.
-    /// Checking logic in this method should be consistent with Ready.containsUpdates().
-    #[inline]
-    pub fn has_ready(&self) -> bool {
-        self.has_ready_since(None)
+    fn commit_ready(&mut self, rd: Ready) {
+        if rd.ss.is_some() {
+            self.prev_ss = rd.ss.unwrap();
+        }
+        if let Some(hs) = rd.hs {
+            if hs != HardState::default() {
+                self.prev_hs = hs;
+            }
+        }
+        let rd_record = self.records.back().unwrap();
+        assert!(rd_record.number == rd.number);
+        if let Some(e) = rd_record.last_entry {
+            self.raft.raft_log.stable_to(e.0, e.1);
+        }
     }
 
-    /// Grabs the snapshot from the raft if available.
-    #[inline]
-    pub fn snap(&self) -> Option<&Snapshot> {
-        self.raft.snap()
+    fn commit_apply(&mut self, applied: u64) {
+        self.raft.commit_apply(applied);
+    }
+
+    /// Notifies that the ready of this number has been well persisted.
+    ///
+    /// Since Ready must be persisted in order, calling this function implicitly means
+    /// all readys with numbers smaller than this have been persisted.
+    pub fn on_persist_ready(&mut self, number: u64) {
+        while let Some(record) = self.records.pop_front() {
+            if record.number > number {
+                break;
+            }
+            if let Some(last_log) = record.last_entry {
+                self.raft.on_persist_entries(last_log.0, last_log.1);
+                self.last_persisted_index = last_log.0;
+            }
+            if !record.snapshot.is_empty() {
+                self.raft
+                    .raft_log
+                    .stable_snap_to(record.snapshot.get_metadata().index);
+                self.pending_snapshot = false;
+            }
+            if !record.messages.is_empty() {
+                self.messages.push(record.messages);
+            }
+        }
+    }
+
+    /// Notifies that the last ready has been well persisted.
+    ///
+    /// Returns the PersistLastReadyResult that contains committed entries and messages.
+    ///
+    /// Since Ready must be persisted in order, calling this function implicitly means
+    /// all readys collected before have been persisted.
+    pub fn on_persist_last_ready(&mut self) -> PersistLastReadyResult {
```

Comment:
As I understand, the async ready convert the process into a pipeline manner with two steps. If there is always getting a new ready(means advancing last ready number), then it hardly can reach the point where the last ready is persisted.

Can we just return `PersistLastReadyResult` for every `on_persist_ready()` and remove `on_persist_last_ready`?

Reply:
My understanding is that these two functions for different purposes.
The `on_persist_ready` aims at batching the message and committed entries. Users can get a Ready after calling several `on_persist_ready` and `step`.
The `on_persist_last_ready` is a fast path for users. If they have already persisted the last Ready, they can get the `PersistLastReadyResult` without getting `Ready` again.

Reply:
`PersistLastReadyResult` seem like the same `Ready` do, can we make `on_persist_last_ready` to be replaced with:
```
self.on_persist_ready(self.max_number);
self.ready()
```

Reply:
Indeed, but I think using `PersistLastReadyResult` instead of `ready`  is better because I want to tell users that **only** committed entries and messages need be handled.

Reply:
>  to tell users that only committed entries and messages are needed to be handled.

I not sure is it better because it forces the user to write some new logic to handle `PersistLastReadyResult` instead of reusing the exiting logic of handling `Ready`

Reply:
I agree with @Connor1996 and @NingLin-P that this extra method make logic complicated and not helpful as it seems.

Reply:
Imagine this case:
1. The application calls several ready and `advance_append_sync`
2. After calling the last ready's `advance_append_sync`, it finds it's time to call fsync
3. If there is no `on_persist_last_ready`, it should call `on_persist_ready` and then call `ready` to get the next ready and the only meaningful things in `Ready` are `committed_entries` and `messages`.

In TiKV, we have lots of logic during handling `Ready` which is very wasteful to get a `ready` again and do the same logic like normal `ready` does just for applying committed entries and sending messages.